33 Fix reliability issues

ja6a opened the pull request.

ja6a committed:
Add go modules

ja6a committed:
initial investigation

ja6a committed:
Add request run state prechecking

ja6a committed:
Cut extra do method

ja6a committed:
Add response checking

ja6a committed:
Add response to review comments

pegerto commented on skytap/client.go:
```diff
@@ -23,9 +26,18 @@ const (
-	defRetryCount = 30
+	defRetryCount = 60
+
+	noRunStateCheck  RunStateCheckStatus = 0
```
is this not better defined as an enum?

ja6a replied:
it already is?

pegerto commented on skytap/client.go:
```diff
@@ -188,40 +212,91 @@ func (c *Client) newRequest(ctx context.Context, method, path string, body inter
-	var resp *http.Response
-	var err error
-	var makeRequest = true
+func (c *Client) do(ctx context.Context, req *http.Request, v interface{}, state *environmentVMRunState, payload responseComparator) (*http.Response, error) {
```
Probably wrap the client in top of a library as [go-retriable](https://godoc.org/github.com/hashicorp/go-retryablehttp#Client) Instead of self implement the logic, it will provide exponential backoff and jitter.

I will probably not differentiate the request by `verb` in do an let the implementation of the check Resource to take that decision.

ja6a replied:
Going to create a new ticket to implement this

pegerto commented on skytap/client.go:
```diff
@@ -287,22 +359,147 @@ func (c *Client) checkResponse(r *http.Response) error {
 	return errorResponse
 }
+
+func (c *Client) checkResourceStateUntilSatisfied(ctx context.Context, req *http.Request, state *environmentVMRunState) error {
```
This logic is very specific for the resource to be implemented at the HTTP client, consider to pass this function as an argument.

ja6a replied:
Going to create a new ticket for this.

pegerto commented on skytap/client.go:
```diff
@@ -188,40 +212,91 @@ func (c *Client) newRequest(ctx context.Context, method, path string, body inter
 
-	return resp, err
+func waitForAwhile(message string, code string, codeAsString string, secondsToWait int, iteration int) {
+	snooze := secondsToWait //* (iteration + 1)
```
I miss the point of this assigment

ja6a replied:
removed

pegerto commented on skytap/client.go:
```diff
@@ -188,40 +212,91 @@ func (c *Client) newRequest(ctx context.Context, method, path string, body inter
-			errBody := readResponseBody(resp, v)
-			if errBody != nil {
-				break
+func (c *Client) requestPutPostDelete(ctx context.Context, req *http.Request, state *environmentVMRunState, payload responseComparator, v interface{}) (*http.Response, bool, error) {
```
you can probably encapsulate 
this method environment method on the response comparator.

ja6a replied:
Going to create a ticket for this.

pegerto commented on skytap/client.go:
```diff
@@ -188,40 +212,91 @@ func (c *Client) newRequest(ctx context.Context, method, path string, body inter
+	var resp *http.Response
+	var err error
+
+	log.Printf("[DEBUG] SDK POST PUT DELETE request (%#v)\n", spew.Sdump(req))
```
The request info already provides the verb, probably can be extracted from the request.

pegerto commented on skytap/client.go:
```diff
@@ -188,40 +212,91 @@ func (c *Client) newRequest(ctx context.Context, method, path string, body inter
+}
 
-	return resp, err
+func waitForAwhile(message string, code string, codeAsString string, secondsToWait int, iteration int) {
```
camel case, probably use the common tern backoff

pegerto commented on skytap/client.go:
```diff
@@ -287,22 +359,147 @@ func (c *Client) checkResponse(r *http.Response) error {
+			} else {
+				ok, err = c.getVMRunState(ctx, state.environmentID, state.vmID, state.vm)
+			}
+			if err != nil {
```
this can be simple if you break the loop in case `err!=nill` or `ok` and return err fewer branches

pegerto commented:
The retry logic makes sense, the main concern is that we make assumptions of if need to be checked, will be nicer if this is pass to the client as a policy.  Consider to leverage solutions as `go-retryable` and the implementation of a specific policy for the skytap client.

ja6a committed:
Fix disk update routine. Fail early on 422 if resource not busy

ja6a commented:
> The retry logic makes sense, the main concern is that we make assumptions of if need to be checked, will be nicer if this is pass to the client as a policy. Consider to leverage solutions as `go-retryable` and the implementation of a specific policy for the skytap client.

I have addressed all the comments:- either directly in the code or by creating new issues to be addressed at a later date. Issues #45 #44

etrueskytap commented on skytap/interface.go:
```diff
@@ -244,10 +248,106 @@ func (s *InterfacesServiceClient) Delete(ctx context.Context, environmentID stri
 	}
 
-	_, err = s.client.do(ctx, req, nil)
+	_, err = s.client.do(ctx, req, nil, nil, nil)
```
Deleting an interface will fail unless the vm is stopped. (Our API docs are wrong here -- I've filed DOC-2954 for correction.)

ja6a replied:
This will be addressed by adding the correct corresponding post request state to the above method on line 251.

etrueskytap commented on skytap/vm.go:
```diff
@@ -458,7 +434,7 @@ func (s *VMsServiceClient) changeRunstate(ctx context.Context, environmentID str
 
 	var updatedVM VM
-	_, err = s.client.do(ctx, requestCreate, &updatedVM)
+	_, err = s.client.do(ctx, requestCreate, &updatedVM, vmRunStateNotBusy(environmentID, id), opts)
```
This needs to check that the environment is not busy. When a VM is busy, its environment is also busy, so attempting to set runstate on other VMs will fail with 423 Locked.

ja6a replied:
Even though the logic as it currently is will work - I will make this change for completeness. In case you were not aware: a 423 error will trigger a retry. Implementing this change will clean up the logs by removing the 423 errors.

etrueskytap commented on skytap/published_service.go:
```diff
@@ -172,10 +177,48 @@ func (s *PublishedServicesServiceClient) Delete(ctx context.Context, environment
+}
+
+func (payload *UpdatePublishedServiceRequest) compare(ctx context.Context, c *Client, v interface{}, state *environmentVMRunState) (string, bool) {
+	return payload.CreatePublishedServiceRequest.compare(ctx, c, v, state)
```
I'm unclear on why we have an `Update` function and an `UpdatePublishedServiceRequest` type. The API only supports create and destroy operations, not update. Do we still need the Update function?

ja6a replied:
Point taken - the function is here to support the terraform provider which requires an 'update' step. Since this is actually a delete and create I will rename the function and corresponding data structures accordingly.

ja6a committed:
Cut Update method and related structures

ja6a committed:
Add requirement for vm to be stopped prior to interface delete

ja6a committed:
Add environment busy check in place of vm ready check

pegerto commented on go.mod:
```diff
@@ -0,0 +1,3 @@
+module github.com/skytap/skytap-sdk-go
```
I will expect this version to be fixed.

pegerto commented on skytap/environment.go:
```diff
@@ -183,7 +186,7 @@ func (s *EnvironmentsServiceClient) List(ctx context.Context) (*EnvironmentListR
 	}
 
 	var environmentsListResponse EnvironmentListResult
-	_, err = s.client.do(ctx, req, &environmentsListResponse.Value)
```
this set of null does not help readability, shall this implement `functional-options` pattern

pegerto commented on skytap/environment.go:
```diff
@@ -217,12 +220,20 @@ func (s *EnvironmentsServiceClient) Create(ctx context.Context, request *CreateE
 
 	var createdEnvironment Environment
-	_, err = s.client.do(ctx, req, &createdEnvironment)
+	_, err = s.client.do(ctx, req, &createdEnvironment, nil, request)
```
haven't read the full code block, but we have `req` and `request` in the same line, this creates ambiguity, shall one of this been renamed?

pegerto commented on skytap/environment.go:
```diff
@@ -234,7 +245,7 @@ func (s *EnvironmentsServiceClient) Create(ctx context.Context, request *CreateE
 		ShutdownOnIdle:  request.ShutdownOnIdle,
 		ShutdownAtTime:  request.ShutdownAtTime,
-		Runstate:        &runstate,
+		Runstate:        runstate,
```
does the logic of starting the environment belong to terraform more than here?  Can we add a comment that explains why the status depends on the environment?

pegerto commented on skytap/environment.go:
```diff
@@ -272,10 +283,150 @@ func (s *EnvironmentsServiceClient) Delete(ctx context.Context, id string) error
 	return nil
 }
+
+func (payload *CreateEnvironmentRequest) compare(ctx context.Context, c *Client, v interface{}, state *environmentVMRunState) (string, bool) {
```
`compare` is so generic  for a name

pegerto commented on skytap/environment.go:
```diff
@@ -272,10 +283,150 @@ func (s *EnvironmentsServiceClient) Delete(ctx context.Context, id string) error
+	return requestNotAsExpected, false
+}
+
+func (payload *UpdateEnvironmentRequest) compare(ctx context.Context, c *Client, v interface{}, state *environmentVMRunState) (string, bool) {
```
should these methods not be part of the environment and not the request, you verify the status, of the instance not the request it self.

ja6a replied:
That could work too...

pegerto commented on skytap/environment.go:
```diff
@@ -272,10 +283,150 @@ func (s *EnvironmentsServiceClient) Delete(ctx context.Context, id string) error
+	return requestNotAsExpected, false
+}
+
+func (payload *UpdateEnvironmentRequest) buildUpdateRequestFromVM(env *Environment) UpdateEnvironmentRequest {
```
why do we need all this if then, can not have nill in the Update Request? why the method name is `fromVM` when retrieves and environment?

ja6a replied:
I think the best answer is because I am using a string to compare the objects.

pegerto commented on skytap/environment.go:
```diff
@@ -272,10 +283,150 @@ func (s *EnvironmentsServiceClient) Delete(ctx context.Context, id string) error
+	if payload.Runstate != nil {
+		runstate = string(*payload.Runstate)
+	}
+	s := fmt.Sprintf("%s%s%s%s%s%s%s%s%s%s",
```
I don't think this is going to be readable if all is place together?

ja6a committed:
Add change to parameter name

ja6a committed:
Add new name for compare

ja6a committed:
Add clarity to method name

ja6a committed:
Add improvement to readability of string formulation

ja6a committed:
Add comment to clarify intention of running the environment
